### PR TITLE
Replace old async loading with new approach

### LINKED ISSUE #2790: Preferences that require a napari restart
The way things are currently implemented, the following will require a restart of napari if they are changed in the Preferences dialog.  Please feel free to add others...

1. octree
2. async

### LINKED ISSUE #5517: Add a public way to enable async slicing
# Description

The `Layer._force_sync` flag effectively acts as a feature/experimental flag that allows to control if async slicing is enabled. If it's true, then we effectively disable async slicing, otherwise we perform it when a layer type supports it.

The goal of this task is to provide at least one public way of controlling that currently private flag. Some possibilities among others.

- Define a napari setting/preference in `NapariSettings`. This will be present in the GUI and will be persisted between napari sessions.
- Define an environment variable name that will be used by napari (e.g. `NAPARI_ASYNC_SLICING`).

One thing that complicates the effort is the [old experimental async implementation](https://napari.org/stable/guides/rendering.html#asynchronous-rendering), which also defines napari settings and environment variables to control whether the async and tiled rendering options are enabled.

We should not try to support or likely even allow the new and old async to be enabled at the same time. Therefore, if both settings are present, we should make them mutually exclusive. Alternatively, we could explore removing the old async style.

# Output

The output of this task should be a PR that adds at least one visible/public way of controlling whether async slicing is enabled or not. This PR should be only be merged after #5236 is done, but could be drafted, prepared, and even reviewed before then.

### PR DESCRIPTION
**Pull request by andy-sweet**
# Fixes/Closes

Closes #5517
Closes #2790

# Description
This replaces [the old approach to asynchronous loading](https://napari.org/0.4.16/guides/rendering.html) with [the new approach described in NAP-4](https://napari.org/stable/naps/4-async-slicing.html).

See the old approach (on `main`) on some remote 3D multi-resolution images from Janelia via [@chili-chiu's fantastic bio sample data plugin](https://www.napari-hub.org/plugins/napari-bio-sample-data):

https://github.com/napari/napari/assets/2608297/e8673d4e-c387-49bc-9254-30d0fc9c1cd2

And the new approach on the same data that is more responsive:

https://github.com/napari/napari/assets/2608297/0b9f1f12-d5e0-4980-be05-c5c58285c1a2

This PR also removes the old approach to tiled rendering (AKA octree) since that is tied to the old approach to asynchronous loading. Most of the code change delta consists of deleted files that are experimental modules and tests associated with those old approaches.

The other changes remove the old integration and adjust the new one to effectively achieve replacement. The key changes are as follows.

1. Remove async/sync_only marks from pytest since these should only be needed for the old approach.
2. Remove command that hides/shows whether octree chunks are visible. I'm unsure if these commands are publicly exposed anywhere, though this one currently appears to be broken.
3. Redefine the existing experimental setting previously used to enable the old approach to enable the new one instead.
4. Add support for `Layer.loaded` for the new approach by adding an ID to each slice request. This could probably be moved to a separate PR if desired (e.g. see https://github.com/andy-sweet/napari/pull/51)
5. Absorb slicing logic spread across old classes into the slice request call.

This PR currently includes #5783, which should be reviewed and merged first to make this large PR as small as possible. For the difference between this and #5783, see https://github.com/andy-sweet/napari/pull/50

# References
I think this improves the simplicity of slicing logic as described in #2156
Also see #5711 for the first version of this PR from @kcpevey, which was split into this and #5783
Part of #4795

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

There probably need to be some documentation updates related to this since the old approach is well documented on napari.org. We should probably write some developer docs based off the async slicing NAP and mark the old docs as deprecated and/or remove them to avoid confusion.

# How has this been tested?
- [x] added an integration test for checking `Layer.loaded` is updated correctly
- [x] all existing tests pass with my change

I also did some manual testing by running `NAPARI_ASYNC=1 napari` and interacting with some remote and non-remote data.

### CODE REVIEW

**Comment by andy-sweet** [comment]
Marking as draft/WIP for now to see if I need to fix any tests.

**Comment by GenevieveBuckley** [comment]
The difference between those two screen recordings is pretty impressive @andy-sweet

**Comment by andy-sweet** [comment]
> The difference between those two screen recordings is pretty impressive @andy-sweet

Thanks! I'm not yet sure why there's such a big difference. My guess is that the old async approach by default has many workers which could cause contention for download bandwidth and the GIL in this single layer case, but may work better for many layers or tiled rendering.

File: napari/components/_tests/test_layer_slicer.py
Comment (kcpevey) [inline napari/components/_tests/test_layer_slicer.py]: Should we have an assert here? We wait for it to get done, but don't check anything after it finishes.

File: napari/components/_tests/test_layer_slicer.py
Comment (kcpevey) [inline napari/components/_tests/test_layer_slicer.py]: Why add these here if we aren't going to use them in testing? I have mixed feelings on this - on one hand I appreciate the Fake Layer being more like a real layer, but on the other, I'd feel more comfortable testing these properties on real code.

File: napari/layers/base/_slice.py
Comment (kcpevey) [inline napari/layers/base/_slice.py]: how does this work?

File: napari/layers/image/_slice.py
Comment (kcpevey) [inline napari/layers/image/_slice.py:248]: NIT: consider moving to an `else` below so we dont create and then sometimes overwrite

File: napari/layers/labels/labels.py
Comment (kcpevey) [inline napari/layers/labels/labels.py:921]: why is this necessary?

**Review by kcpevey — APPROVED** [review]
Looks good. I left a few lingering comments after the latest updates. Nothing major though.

File: napari/components/_tests/test_layer_slicer.py
Comment (kcpevey) [reply napari/components/_tests/test_layer_slicer.py]: And also... there are no tests here using these two methods.

Comment (kcpevey) [reply napari/components/_tests/test_layer_slicer.py]: `_set_unloaded_slice_id` is now part of the async sliceable layer protocol

File: napari/layers/labels/labels.py
Comment (andy-sweet) [reply napari/layers/labels/labels.py:921]: We discussed this over zoom. I think the docstring does a decent job of explaining why, but other suggestions welcome. I also considered moving the body of `to_displayed` here since this is the only place this is needed, but I will probably hold off on that for that to avoid some extra imports.

File: napari/layers/base/_slice.py
Comment (andy-sweet) [reply napari/layers/base/_slice.py]: We discussed over zoom and I added a few more docs here to help clarify.

File: napari/components/_tests/test_layer_slicer.py
Comment (andy-sweet) [reply napari/components/_tests/test_layer_slicer.py]: I removed `_set_loaded` for now since that is not needed. We also discussed using real layers now to avoid these fakes, which could be doable either as part of a another separate PR like https://github.com/andy-sweet/napari/pull/51 or in a follow-up PR.

File: napari/components/_tests/test_layer_slicer.py
Comment (andy-sweet) [reply napari/components/_tests/test_layer_slicer.py]: Since this isn't really relevant to the changes in this PR and we have an integration test that checks the sliced points data, I removed this.

**Comment by brisvag** [comment]
Awesome work! Always good to see net improvement paired with big red numbers in the diff :) I'll take a deeper look soon, for now jus tbig excitement!

File: napari/_qt/qt_viewer.py
Comment (andy-sweet) [inline napari/_qt/qt_viewer.py:597]: @jni : would be useful to get your feedback on the ordering here (i.e. update the `loaded` state after everything else slice related), since this call may also trigger `Layer.events.loaded` to be emitted. That's slightly different to the old async approach where `loaded` is effectively updated in the equivalent of `_update_slice_response` (so that the vispy image code can use the value of `loaded` appropriately).

File: napari/_qt/_qapp_model/qactions/_view.py
Comment (andy-sweet) [inline napari/_qt/_qapp_model/qactions/_view.py:82]: Unsure if it's safe to just remove an action. Is there a deprecation path to follow here?

**Comment by andy-sweet** [comment]
There is a bug related to changing `Dims.order` for multiscale images when async is on, for which I have a [seemingly reliable reproducer](https://gist.github.com/andy-sweet/55549cb1872383c8c78e949ecde3e246), though there may be some napari canvas / screen resolution dependence. I also have some [extra logs to understand when this is occurring](https://gist.github.com/andy-sweet/f2f53bed527188caded33cf67d8b7261).

The exception is actually converted to a warning by vispy and eventually the desired slice is drawn, so it's not too severe. But the napari notification dialog pops up, which is disappointing. I'm suspicious of `Layer.corner_pixels` (maybe because I always am!), but I haven't made too much solid progress or found any possible leads in terms of fixes.

Possibly related to https://github.com/napari/napari/issues/5106 or maybe https://github.com/napari/napari/issues/5409

Could definitely do with some help from vispy experts if they have time: @brisvag or @aganders3 maybe?

**Comment by andy-sweet** [comment]
> There is a bug related to changing `Dims.order` for multiscale images when async is on, for which I have a [seemingly reliable reproducer](https://gist.github.com/andy-sweet/55549cb1872383c8c78e949ecde3e246), though there may be some napari canvas / screen resolution dependence. I also have some [extra logs to understand when this is occurring](https://gist.github.com/andy-sweet/f2f53bed527188caded33cf67d8b7261).
> 
> The exception is actually converted to a warning by vispy and eventually the desired slice is drawn, so it's not too severe. But the napari notification dialog pops up, which is disappointing. I'm suspicious of `Layer.corner_pixels` (maybe because I always am!), but I haven't made too much solid progress or found any possible leads in terms of fixes.
> 
> Possibly related to #5106 or maybe #5409
> 
> Could definitely do with some help from vispy experts if they have time: @brisvag or @aganders3 maybe?

After a little more digging, I'm pretty confident the bug is caused by [`corner_pixels` since it is updated in `Layer._update_draw` based on `Layer._slice_input`](https://github.com/napari/napari/blob/0dfb37b474490b4d9829c4fa93cb3d648b4b186d/napari/layers/base/base.py#L1634). With the new async that update is temporarily incorrect after a change to `order` since `Layer._slice_input` is only updated as state once the output/response is ready - so that those two can be queried in sync. The old async also kept them in sync but by updating `_slice_input` when slicing starts and wiping out the last output/response to be empty, so managed to avoid this particular issue, but had other undesirable consequences (e.g. `get_value` returns nothing when inspecting images displayed on the canvas).

My rough fix (which works in this case) is to pass through `displayed_axes` from [`VispyCanvas.on_draw` but even that is partially dependent on `Layer._slice_input` sometimes](https://github.com/napari/napari/blob/0dfb37b474490b4d9829c4fa93cb3d648b4b186d/napari/_vispy/canvas.py#LL514C19-L514C19). I think about ways to improve/simplify the code in `VispyCanvas.on_draw` that also prevents this bug in general.

**Comment by andy-sweet** [comment]
Just chatted with @aganders3 more about the `order` bug. We thought of some other potential fixes.

1. Use the world-to-layer correspondence code to replace some of the code in `VispyCanvas.on_draw`.
2. Skip empty (or maybe < 2D) data in the vispy image data.
3. Compute `corner_pixels` before drawing (e.g. by connecting to vispy camera events).
4. Store the last pending slice input, since we have the last complete slice input in the response.

To start I'm going to look into (1) and (2), which might also fix related sync-bugs like #5106. I'm going to avoid (4) for now, since that may require quite a lot of updates throughout the code base since there is a lot of dependence on `Layer._slice_input` right now. (3) is independently useful, so is useful future work too.

**Comment by andy-sweet** [comment]
Also, regardless of the order bug (which is not severe), this PR is ready for review. Would be great to get at least a couple of written reviews from @napari/core-devs given the gravity of the changes here. I've walked through this with a few people, but am happy to schedule more time for interactive reviews - just send me a direct message on Zulip to set up a time.

**Comment by brisvag** [comment]
Note that after merging #5751 the whole corner_pixel situation needs to be reevaluated!

File: napari/_qt/_qapp_model/qactions/_view.py
Comment (brisvag) [reply napari/_qt/_qapp_model/qactions/_view.py:82]: I would say that at least for actions related to experimental features we can just remove them without deprecation.

**Review by brisvag — APPROVED** [review]
Finally got around to looking at this a bit more thoroughly. Awesome work @andy-sweet and @kcpevey! As far as I can tell, everything makes sense to me. Maybe better to get an extra opinion, but other than that not much else to say other than let's get this in asap :D

**Comment by andy-sweet** [comment]
Since this is over 1 month old, has two approvals, one form a core dev that's over a week old, I'll likely merge this by the middle or end of next week unless there are any objections. Feedback is still welcome though!

File: napari/_qt/qt_viewer.py
Comment (jni) [reply napari/_qt/qt_viewer.py:597]: I don't understand the full consequences of your question... If the VisPy layer no longer needs that value (?), then yes, doing it at the end seems like the most intuitive thing...

File: napari/components/_layer_slicer.py
Comment (jni) [inline napari/components/_layer_slicer.py]: I feel like this warrants a comment about why it exists... 😅

File: napari/layers/labels/labels.py
Comment (jni) [reply napari/layers/labels/labels.py:921]: Yeah, sorry about this. 😂 Seriously I can remember thinking, "let's just get a hacky implementation of labels going that just reuses image, we can tidy it up later." This is like Feb 2019. 😂  #3308 will slightly improve the situation but it'll still be an `astype` call...

**Review by jni — APPROVED** [review]

File: napari/components/_layer_slicer.py
Comment (andy-sweet) [reply napari/components/_layer_slicer.py]: Good idea. I added a few docstrings to provide more context for this protocol and its methods.

**Pull request merged**
